monix-tail/shared/src/main/scala/monix/tail/internal/IterantOnError.scala

```diff
@@ -48,18 +49,23 @@ private[tail] object IterantOnError {
       Suspend[F, A](next, stop)
     }
 
+    def tailGuard(stop: F[Unit])(ffa: F[Iterant[F, A]]): F[Iterant[F, A]] =
+      ffa.handleError(sendError(stop, _))
+
     def loop(fa: Iterant[F, A]): Iterant[F, A] =
       try fa match {
         case Next(a, lt, stop) =>
-          Next(a, lt.map(loop), stop)
+          Next(a, tailGuard(stop)(lt.map(loop)), stop)
```

Comment:
Everything looks good except that I feel this call is overreaching. It should probably not include the `.map(loop)`:

```scala
tailGuard(lt, stop).map(loop)
```

For one this is because it's unnecessary and also because we're avoiding optimization opportunities — e.g. `lt` could be a `pure` value for which the `F[_]` implementation could bypass error handling.